packages/plugin-workspaces-tools/sources/commands/foreach.ts

```diff
@@ -0,0 +1,187 @@
+    .describe(`run a command on all workspaces`)
+
+    .action(
+      async ({cwd, args, stdout, command, exclude, include, interlaced, parallel, withDependencies, prefixed, ...env}: ForeachOptions) => {
+        const configuration = await Configuration.find(cwd, pluginConfiguration);
+        const { project } = await Project.find(configuration, cwd);
+        const cache = await Cache.find(configuration);
+
+        const needsProcessing = new Map<LocatorHash, Workspace>();
+        const processing = new Set<DescriptorHash>();
+        const concurrency = parallel ? Math.max(1, cpus().length / 2) : 1;
+        const limit = pLimit(concurrency);
+        let commandCount = 0;
+
+        const report = await StreamReport.start({configuration, stdout}, async report => {
+          await project.resolveEverything({lockfileOnly: true, cache, report});
+
+          let workspaces = command.toLowerCase() === `run`
+            ? project.workspaces.filter(workspace => workspace.manifest.scripts.has(args[0]))
+            : project.workspaces;
+
+          if (include.length > 0) {
+            workspaces = workspaces.filter(workspace => include.includes(workspace.locator.name))
+          }
+
+          if (exclude.length > 0) {
+            workspaces = workspaces.filter(workspace => !exclude.includes(workspace.locator.name))
+          }
+
+          for (const workspace of workspaces) {
+            needsProcessing.set(workspace.locator.locatorHash, workspace);
+          }
+
+          while (needsProcessing.size > 0) {
+            const commandPromises = [];
+
+            for (const [identHash, workspace] of needsProcessing) {
+              // If we are already running the command on that workspace, skip
+              if (processing.has(workspace.anchoredDescriptor.descriptorHash)) {
+                continue;
+              }
+
+              let isRunnable = true;
+
+              // By default we do topological, however we don't care of the order when running
+              // in --parallel unless also given the --with-dependencies flag
+              if (!parallel || withDependencies) {
+                for (const [identHash, descriptor] of workspace.dependencies) {
+                  const locatorHash = project.storedResolutions.get(descriptor.descriptorHash);
+
+                  if (needsProcessing.has(locatorHash!))
+                    isRunnable = false;
+                }
+              }
+
+              if (!isRunnable) {
+                continue;
+              }
+
+              processing.add(workspace.anchoredDescriptor.descriptorHash);
+
+              commandPromises.push(limit(async () => {
+                commandCount += 1;
+                await runCommand(workspace);
+                needsProcessing.delete(identHash);
+                processing.delete(workspace.anchoredDescriptor.descriptorHash);
+              }));
+            }
+
+            await Promise.all(commandPromises);
```

Comment:
If there's a loop in the workspace's dependencies, we'll loop forever. Throwing an error if `commandPromises` is empty would catch this situation

Reply:
Good catch 💯